libs/service/api_service.cpp

```diff
@@ -17,7 +16,181 @@ extern "C"
 }
 #include "Verifier.h"
 #include "verifier_service.h"
-#include "windows_helpers.hpp"
+
+#define MAX_CODE_SIZE (32 * 1024) // 32 KB
+
+static ebpf_result_t
+_build_helper_id_to_address_map(
+    ebpf_handle_t program_handle, ebpf_code_buffer_t& byte_code, std::map<uint32_t, uint64_t>& helper_id_to_adddress)
+{
+    ebpf_inst* instructions = reinterpret_cast<ebpf_inst*>(byte_code.data());
+    for (size_t index = 0; index < byte_code.size() / sizeof(ebpf_inst); index++) {
+        ebpf_inst& instruction = instructions[index];
+        if (instruction.opcode != INST_OP_CALL) {
+            continue;
+        }
+        helper_id_to_adddress[instruction.imm] = 0;
+    }
+
+    if (helper_id_to_adddress.size() == 0)
+        return EBPF_SUCCESS;
+
+    ebpf_protocol_buffer_t request_buffer(
+        offsetof(ebpf_operation_resolve_helper_request_t, helper_id) + sizeof(uint32_t) * helper_id_to_adddress.size());
+
+    ebpf_protocol_buffer_t reply_buffer(
+        offsetof(ebpf_operation_resolve_helper_reply_t, address) + sizeof(uint64_t) * helper_id_to_adddress.size());
+
+    auto request = reinterpret_cast<ebpf_operation_resolve_helper_request_t*>(request_buffer.data());
+    auto reply = reinterpret_cast<ebpf_operation_resolve_helper_reply_t*>(reply_buffer.data());
+    request->header.id = ebpf_operation_id_t::EBPF_OPERATION_RESOLVE_HELPER;
+    request->header.length = static_cast<uint16_t>(request_buffer.size());
+    request->program_handle = reinterpret_cast<uint64_t>(program_handle);
+
+    size_t index = 0;
+    for (const auto& helper : helper_id_to_adddress) {
+        request->helper_id[index++] = helper.first;
+    }
+
+    uint32_t result = invoke_ioctl(device_handle, request_buffer, reply_buffer);
+    if (result != ERROR_SUCCESS) {
+        return windows_error_to_ebpf_result(result);
+    }
+
+    index = 0;
+    for (auto& helper : helper_id_to_adddress) {
+        helper.second = reply->address[index++];
+    }
+
+    return EBPF_SUCCESS;
+}
+
+static ebpf_result_t
+_resolve_ec_function(ebpf_ec_function_t function, uint64_t* address)
+{
+    ebpf_operation_get_ec_function_request_t request = {sizeof(request), EBPF_OPERATION_GET_EC_FUNCTION, function};
+    ebpf_operation_get_ec_function_reply_t reply;
+
+    uint32_t result = invoke_ioctl(device_handle, request, reply);
+    if (result != ERROR_SUCCESS) {
+        return windows_error_to_ebpf_result(result);
```

Comment:
move windows_error_to_ebpf_result call inside the invoke_ioctl() method

Reply:
I think same error code from invoke_ioctl may need to be converted to a different ebpf_result, depending on the situation.

Example:
When invoke_ioctl fails in case of EBPF_OPERATION_LOAD_CODE, "EBPF_PROGRAM_LOAD_FAILED" is returned.
When invoke_ioctl fails in case of any map operation, the app will get some different ebpf_result error code (like KEY_NOT_FOUND)?

Reply:
Ok that's an acceptable answer as far as invoke_ioctl returning a win32 error code.  But then line 76 would be wrong because it doesn't take the request type into account.

Reply:
In this specific case (_resolve_ec_function), it is ok to directly convert the win32 error code to ebpf result (and not do any special conversion).

Reply:
If you think so, yes.